worker-store-http/src/main/java/com/hpe/caf/worker/datastore/http/HttpDataStore.java

```diff
@@ -0,0 +1,327 @@
+                throw new DataStoreException(String.format(
+                    "Failed to store data with reference: %s. Response code: %s. Response message: %s",
+                    reference, httpUrlConnection.getResponseCode(), httpUrlConnection.getResponseMessage()));
+            }
+        } catch (final IOException e) {
+            numErrors.incrementAndGet();
+            throw new DataStoreException(String.format("Failed to store data with reference: %s", reference), e);
+        } finally {
+            if (dataOutputStream != null) {
+                try {
+                    dataOutputStream.flush();
+                    dataOutputStream.close();
+                } catch (final IOException e) {
+                    LOG.warn("Unable to flush or close output stream", e);
+                }
+            }
+            if (httpUrlConnection != null) {
+                httpUrlConnection.disconnect();
+            }
+        }
+    }
+
+    @Override
+    public String store(final InputStream inputStream, final String partialReference)
+        throws DataStoreException
+    {
+        try {
+            return store(ByteStreams.toByteArray(inputStream), partialReference);
+        } catch (IOException e) {
+            throw new DataStoreException(String.format(
+                "Failed to store data with  partial reference (may be null, which is allowed): %s, "
+                + "as unable to create byte array from input stream.", partialReference), e);
+        }
+    }
+
+    @Override
+    public String store(final Path path, final String partialReference)
+        throws DataStoreException
+    {
+        try (QuietResource<FileInputStream> inputStream = new QuietResource<>(new FileInputStream(path.toFile()))) {
+            return store(inputStream.get(), partialReference);
+        } catch (final IOException ex) {
+            numErrors.incrementAndGet();
+            throw new DataStoreException(String.format(
+                "Unable to store data withCould not create file input stream from path: %s.", path.toString()), ex);
+        }
+    }
+
+    private class HttpDataStoreMetricsReporter implements DataStoreMetricsReporter
+    {
+        @Override
+        public int getDeleteRequests()
+        {
+            return numDeleteRequests.get();
+        }
+
+        @Override
+        public int getStoreRequests()
+        {
+            return numStoreRequests.get();
+        }
+
+        @Override
+        public int getRetrieveRequests()
+        {
+            return numRetrieveRequests.get();
+        }
+
+        @Override
+        public int getErrors()
+        {
+            return numErrors.get();
+        }
+    }
+
+    private static int getConnectTimeoutMillis(final HttpDataStoreConfiguration config)
+    {
+        final Integer configValue = config.getConnectTimeoutMillis();
+        return (configValue == null || configValue <= 0) ? DEFAULT_CONNECT_TIMEOUT_MILLIS : configValue;
+    }
```

Comment:
I think there may be some reason why we don't do this. I followed the pattern of how the `FileSystemDataStore` handles config, which uses a boxed Integer, and does validation checks in code rather than relying on annotations:

https://github.com/WorkerFramework/worker-framework/commit/71806a0aa9e85b84adcca9701c4c6d1854aa9f97#diff-fe413f42103b729013671672738ab1e5ba59967579966aa46cde14c8f039dcb1R348

```
private static int getOutputBufferSize(final FileSystemDataStoreConfiguration config)
    {
        final int DEFAULT_OUTPUT_BUFFER_SIZE = 16384;

        final Integer configSetting = config.getOutputBufferSize();

        return (configSetting == null || configSetting <= 0)
            ? DEFAULT_OUTPUT_BUFFER_SIZE
            : configSetting;
    }
```

It may be that the annotations don't get checked in some scenarios, i.e. usingthe example below, I create a new config object and set `connectTimeoutMillis` to -1, and the annotations are not checked, whereas with the above example, the checks in the code prevent getting a value <=0 
```
    @Min(1)
    private Integer connectTimeoutMillis;

    @SuppressWarnings("null") // Not unboxing null value, checked before returning
    public Integer getConnectTimeoutMillis()
    {
        return connectTimeoutMillis != null ? connectTimeoutMillis : DEFAULT_CONNECT_TIMEOUT_MILLIS;
    }
```

But I'm not entirely sure, maybe @dermot-hardy may be able to provide more info on it?

Reply:
Looking at this again, I think this is how it works:

1) The url property is annotated in the config because it is a required property, so the annotations should always be checked

2) The non-required, numeric properties, such as the timeouts, are not annotated with constraints because it's possible that they may not be provided. If they are not provided, we want to use default values. A boxed Integer is used so we can use null to check for the absence of the value, rather than using 0 if we used a primitve int, which may be confusing.

Having said that, I think the code and constants relating to the checking can move to the config object as you say, so have moved them now.